File: frontend/src/types/studyPlan.ts
/**
 * TypeScript types for Study Plans feature.
 * Matches backend Pydantic models in backend/app/models/study_plan.py
 */

// ============================================
// Enums
// ============================================

export type PlanStatus = 'active' | 'paused' | 'completed';

export type TaskType = 'practice_questions' | 'review_bookmarks' | 'review_srs' | 'study_category';

export type TaskStatus = 'pending' | 'completed' | 'skipped';

// ============================================
// Study Plan Settings
// ============================================

export interface StudyPlanSettings {
    daily_question_goal: number;
    categories: string[];
    difficulty_mix: Record<string, number>;
    include_srs_reviews: boolean;
    include_bookmark_reviews: boolean;
    reminder_time?: string;
    study_days: number[]; // 1=Monday, 7=Sunday
}

export const DEFAULT_SETTINGS: StudyPlanSettings = {
    daily_question_goal: 10,
    categories: [],
    difficulty_mix: { easy: 0.2, medium: 0.5, hard: 0.3 },
    include_srs_reviews: true,
    include_bookmark_reviews: true,
    reminder_time: undefined,
    study_days: [1, 2, 3, 4, 5, 6, 7],
};

// ============================================
// Milestone Types
// ============================================

export interface MilestoneCreate {
    title: string;
    description?: string;
    target_date?: string;
    order_index: number;
}

export interface MilestoneUpdate {
    title?: string;
    description?: string;
    target_date?: string;
}

export interface Milestone extends MilestoneCreate {
    id: string;
    plan_id: string;
    completed_at?: string;
    created_at: string;
}

// ============================================
// Daily Task Types
// ============================================

export interface DailyTaskCreate {
    date: string;
    task_type: TaskType;
    target_count?: number;
}

export interface DailyTaskUpdate {
    completed_count?: number;
    status?: TaskStatus;
}

export interface DailyTask extends DailyTaskCreate {
    id: string;
    plan_id: string;
    completed_count: number;
    status: TaskStatus;
    created_at: string;
}

// ============================================
// Study Plan Types
// ============================================

export interface StudyPlanCreate {
    name: string;
    description?: string;
    target_date?: string;
    settings?: Partial<StudyPlanSettings>;
    milestones?: MilestoneCreate[];
}

export interface StudyPlanUpdate {
    name?: string;
    description?: string;
    target_date?: string;
    settings?: Partial<StudyPlanSettings>;
    status?: PlanStatus;
}

export interface StudyPlan {
    id: string;
    user_id: string;
    name: string;
    description?: string;
    target_date?: string;
    status: PlanStatus;
    settings: StudyPlanSettings;
    milestones: Milestone[];
    created_at: string;
    updated_at: string;
}

export interface StudyPlanSummary {
    id: string;
    user_id: string;
    name: string;
    description?: string;
    target_date?: string;
    status: PlanStatus;
    created_at: string;
    updated_at: string;
}

// ============================================
// Progress Types
// ============================================

export interface StudyPlanProgress {
    plan_id: string;
    total_milestones: number;
    completed_milestones: number;
    total_tasks: number;
    completed_tasks: number;
    current_streak: number;
    overall_progress: number; // 0-100 percentage
    days_remaining?: number;
    estimated_completion?: string;
}

export interface TodayTasksSummary {
    total_tasks: number;
    completed_tasks: number;
    total_questions_target: number;
    total_questions_completed: number;
    plans: Array<{
        plan_id: string;
        plan_name: string;
        tasks: Array<{
            id: string;
            type: TaskType;
            target?: number;
            completed: number;
            status: TaskStatus;
        }>;
    }>;
}

// ============================================
// Calendar Types
// ============================================

export interface CalendarDay {
    date: string;
    tasks: DailyTask[];
    total_target: number;
    total_completed: number;
    status: 'empty' | 'pending' | 'partial' | 'complete';
}

export interface CalendarMonth {
    year: number;
    month: number;
    days: CalendarDay[];
}

// ============================================
// Task Generation Request
// ============================================

export interface TaskGenerationRequest {
    start_date: string;
    end_date: string;
    overwrite_existing?: boolean;
}

// ============================================
// API Response Types
// ============================================

export interface StudyPlanListResponse {
    plans: StudyPlanSummary[];
    total: number;
}

export interface DeleteResponse {
    message: string;
}

File: backend/app/models/study_plan.py
"""
Pydantic models for Study Plans API.
Includes models for plans, milestones, and daily tasks.
"""
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from datetime import date, datetime
from enum import Enum


class PlanStatus(str, Enum):
    """Study plan status options."""
    ACTIVE = "active"
    PAUSED = "paused"
    COMPLETED = "completed"


class TaskType(str, Enum):
    """Types of daily tasks."""
    PRACTICE_QUESTIONS = "practice_questions"
    REVIEW_BOOKMARKS = "review_bookmarks"
    REVIEW_SRS = "review_srs"
    STUDY_CATEGORY = "study_category"


class TaskStatus(str, Enum):
    """Daily task status options."""
    PENDING = "pending"
    COMPLETED = "completed"
    SKIPPED = "skipped"


# ============================================
# Study Plan Settings
# ============================================

class StudyPlanSettings(BaseModel):
    """Settings for a study plan."""
    daily_question_goal: int = Field(
        default=10,
        ge=1,
        le=100,
        description="Number of questions to practice daily"
    )
    categories: List[str] = Field(
        default_factory=list,
        description="Categories to focus on in this plan"
    )
    difficulty_mix: Dict[str, float] = Field(
        default={"easy": 0.2, "medium": 0.5, "hard": 0.3},
        description="Distribution of question difficulties"
    )
    include_srs_reviews: bool = Field(
        default=True,
        description="Include spaced repetition reviews in daily tasks"
    )
    include_bookmark_reviews: bool = Field(
        default=True,
        description="Include bookmark reviews in daily tasks"
    )
    reminder_time: Optional[str] = Field(
        default=None,
        description="Daily reminder time in HH:MM format"
    )
    study_days: List[int] = Field(
        default=[1, 2, 3, 4, 5, 6, 7],
        description="Days of week to study (1=Monday, 7=Sunday)"
    )


# ============================================
# Milestone Models
# ============================================

class MilestoneCreate(BaseModel):
    """Model for creating a milestone."""
    title: str = Field(..., max_length=200, description="Milestone title")
    description: Optional[str] = Field(default=None, description="Milestone description")
    target_date: Optional[date] = Field(default=None, description="Target completion date")
    order_index: int = Field(..., ge=0, description="Order position in the plan")


class MilestoneUpdate(BaseModel):
    """Model for updating a milestone."""
    title: Optional[str] = Field(default=None, max_length=200)
    description: Optional[str] = None
    target_date: Optional[date] = None


class Milestone(MilestoneCreate):
    """Complete milestone model."""
    id: str
    plan_id: str
    completed_at: Optional[datetime] = None
    created_at: datetime


# ============================================
# Daily Task Models
# ============================================

class DailyTaskCreate(BaseModel):
    """Model for creating a daily task."""
    date: date
    task_type: TaskType
    target_count: Optional[int] = Field(default=None, ge=1)


class DailyTaskUpdate(BaseModel):
    """Model for updating a daily task."""
    completed_count: Optional[int] = Field(default=None, ge=0)
    status: Optional[TaskStatus] = None


class DailyTask(DailyTaskCreate):
    """Complete daily task model."""
    id: str
    plan_id: str
    completed_count: int = 0
    status: TaskStatus = TaskStatus.PENDING
    created_at: datetime


# ============================================
# Study Plan Models
# ============================================

class StudyPlanCreate(BaseModel):
    """Model for creating a study plan."""
    name: str = Field(..., max_length=200, description="Plan name")
    description: Optional[str] = Field(default=None, description="Plan description")
    target_date: Optional[date] = Field(default=None, description="Goal completion date")
    settings: StudyPlanSettings = Field(default_factory=StudyPlanSettings)
    milestones: List[MilestoneCreate] = Field(default_factory=list)


class StudyPlanUpdate(BaseModel):
    """Model for updating a study plan."""
    name: Optional[str] = Field(default=None, max_length=200)
    description: Optional[str] = None
    target_date: Optional[date] = None
    settings: Optional[StudyPlanSettings] = None
    status: Optional[PlanStatus] = None


class StudyPlan(BaseModel):
    """Complete study plan model."""
    id: str
    user_id: str
    name: str
    description: Optional[str] = None
    target_date: Optional[date] = None
    status: PlanStatus = PlanStatus.ACTIVE
    settings: StudyPlanSettings
    milestones: List[Milestone] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime


class StudyPlanSummary(BaseModel):
    """Summary model for list views (without milestones)."""
    id: str
    user_id: str
    name: str
    description: Optional[str] = None
    target_date: Optional[date] = None
    status: PlanStatus = PlanStatus.ACTIVE
    created_at: datetime
    updated_at: datetime


# ============================================
# Progress Models
# ============================================

class StudyPlanProgress(BaseModel):
    """Progress summary for a study plan."""
    plan_id: str
    total_milestones: int
    completed_milestones: int
    total_tasks: int
    completed_tasks: int
    current_streak: int
    overall_progress: float = Field(..., ge=0, le=100, description="Progress percentage 0-100")
    days_remaining: Optional[int] = None
    estimated_completion: Optional[date] = None


class TodayTasksSummary(BaseModel):
    """Summary of today's tasks across all plans."""
    total_tasks: int
    completed_tasks: int
    total_questions_target: int
    total_questions_completed: int
    plans: List[Dict[str, Any]] = Field(default_factory=list)


# ============================================
# Calendar Models
# ============================================

class CalendarDay(BaseModel):
    """Single day in calendar view."""
    date: date
    tasks: List[DailyTask] = Field(default_factory=list)
    total_target: int = 0
    total_completed: int = 0
    status: str = "empty"  # empty, pending, partial, complete


class CalendarMonth(BaseModel):
    """Month view for calendar."""
    year: int
    month: int
    days: List[CalendarDay]


# ============================================
# Task Generation Request
# ============================================

class TaskGenerationRequest(BaseModel):
    """Request to generate tasks for a date range."""
    start_date: date
    end_date: date
    overwrite_existing: bool = Field(default=False)
